Fixed slider value indicator not disappearing after a bit on desktop platform when slider is clicked not dragged

cruiser-baxter opened the pull request:
In slider.dart within the _startInteraction method and within the below conditional.

"if (!_active && !hasFocus &&
_state.valueIndicatorController.status == AnimationStatus.completed)"

**Changed to:**

"f (!_active &&
_state.valueIndicatorController.status == AnimationStatus.completed)"
This allows the value indicator to disappear after a bit when clicked instead of dragged on Desktop platform. 

I also added a test in slider_test.dart to detect the bug if it ever returns.

Fixes https://github.com/flutter/flutter/issues/123313

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

cruiser-baxter committed:
slider_test.dart added a test for the desktop slider to check that when the slider is clicked and not dragged, the slider value indicator dissappears after a bit

cruiser-baxter committed:
Reformatted the desktop platform slider value indicator disappears after a bit when clicked test and added some comments

cruiser-baxter committed:
In the _startInteraction method, changed the if(NOT _active && NOT hasFocus && _state.valueIndicatorController.status == AnimationStatus.completed) conditional check, removed the && NOT hasFocus check.

cruiser-baxter committed:
Merge branch 'flutter:master' into b123313

cruiser-baxter committed:
Merge branch 'flutter:master' into b123313

cruiser-baxter committed:
renamed test title to conform with a similar existing test. Removed duplicate pumpAndSettle call before the timed call.

cruiser-baxter committed:
Merge branch 'master' into b123313

TahaTesser reviewed (changes requested):
Thanks for fixing this! 
A few changes are needed then it should be good to go.

TahaTesser reviewed (changes requested):
Thanks for fixing this! 
A few changes are needed then it should be good to go.

cruiser-baxter committed:
Merge branch 'master' into b123313

cruiser-baxter committed:
Update packages/flutter/lib/src/material/slider.dart

Committing changes suggested by TahaTesser. Moved 
 if conditional to single line.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter committed:
Update packages/flutter/test/material/slider_test.dart

COmmitting changes suggested by TahaTesser. Significantly simplified the test, updated test title, and added a comment indicating that it is a regression test for buf #123313.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter commented:
@TahaTesser, I sincerely appreciate your review and suggestions. After going over the diff for my own learning I committed the suggested changes. I learned a lot from the simplified test. Thank you.

TahaTesser reviewed (approved):
LGTM!

TahaTesser commented:
This will need a second review from @Piinks to land.

Piinks reviewed (commented):
Hey @cruiser-baxter, welcome! Thanks for contributing!
The description and test sound like this is specific to desktop, does this change alter the behavior on mobile? Will the label disappear now on mobile if there is focus? Might be good to add a test for this (if one does not already exist)
Looks like the `!hasFocus` was added in https://github.com/flutter/flutter/pull/113543 by @TahaTesser, which was also desktop specific.

TahaTesser commented:
> Hey @cruiser-baxter, welcome! Thanks for contributing! The description and test sound like this is specific to desktop, does this change alter the behavior on mobile? 

This doesn't change the behavior on mobile. `_endInteraction` always dismisses the value indicator. 


> Looks like the !hasFocus was added in https://github.com/flutter/flutter/pull/113543 by @TahaTesser, which was also desktop specific.

Indeed I added it but based on the issue report, I found that interacting directly should dismiss the value indicator.  This issue fixes that interaction behavior.

While on the desktop the value indicator appears but doesn't dismiss when only using the keyboard. 
However current Slider implementation doesn't support invoking value indicators this way.  https://github.com/flutter/flutter/issues/113538. 

> Might be good to add a test for this (if one does not already exist)

Actually, we can remove the desktop target on this test. Since the interaction behavior is the same on desktop and mobile.

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,54 @@ void main() {
+      valueIndicatorBox,
+      isNot(paints..path(color: const Color(0xff000000))..paragraph()),
+    );
+  }, variant: TargetPlatformVariant.desktop());
```
We don't need this.

```suggestion
  });
```

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,54 @@ void main() {
   }, variant: TargetPlatformVariant.desktop());
 
+  testWidgets('Value indicator disappears after adjusting the slider', (WidgetTester tester) async {
+    // This is a regression test for https://github.com/flutter/flutter/issues/123313.
+    const double currentValue = 0.5;
+    await tester.pumpWidget(MaterialApp(
+      home: Material(
```
To make this test future-proof. Let's Material 3 color to check value indicator visibility.
We won't need to update this later. 

```suggestion
    // This is a regression test for https://github.com/flutter/flutter/issues/123313.
    final ThemeData theme = ThemeData(useMaterial3: true);
    const double currentValue = 0.5;
    await tester.pumpWidget(MaterialApp(
      theme: theme,
      home: Material(
```

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,54 @@ void main() {
+    RenderBox valueIndicatorBox = tester.renderObject(find.byType(Overlay));
+    expect(
+      valueIndicatorBox,
+      isNot(paints..path(color: const Color(0xff000000))..paragraph()),
```
```suggestion
      isNot(paints..path(color: theme.colorScheme.primary)..paragraph()),
```

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,54 @@ void main() {
+    valueIndicatorBox = tester.renderObject(find.byType(Overlay));
+    expect(
+      valueIndicatorBox,
+      paints..path(color: const Color(0xff000000))..paragraph(),
```
```suggestion
      paints..path(color: theme.colorScheme.primary)..paragraph(),
```

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,54 @@ void main() {
+    // Value indicator is no longer visible.
+    expect(
+      valueIndicatorBox,
+      isNot(paints..path(color: const Color(0xff000000))..paragraph()),
```
```suggestion
      isNot(paints..path(color: theme.colorScheme.primary)..paragraph()),
```

cruiser-baxter committed:
Update packages/flutter/test/material/slider_test.dart

Committing changes suggested by TahaTesser. Change removes desktop specific platform target.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter committed:
Update packages/flutter/test/material/slider_test.dart

Committing changes suggested by TahaTesser. Changes make the test future proof by using material 3 color.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter committed:
Update packages/flutter/test/material/slider_test.dart

Committing changes suggested by TahaTesser. Changes the color being checcked for  from hex to material colorScheme.primary.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter committed:
Update packages/flutter/test/material/slider_test.dart

Committing changes suggested by TahaTesser. Changes color being checked for from hex to colorScheme.primary.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter committed:
Update packages/flutter/test/material/slider_test.dart

Committing changes suggested by TahaTesser. Changes color being checked for from hex to colorScheme.primary.

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

cruiser-baxter committed:
Merge branch 'master' into b123313

cruiser-baxter reviewed (commented):
I have reviewed the suggested changes and committed all suggestions. Thank you TahaTesser for all of your input, feedback, and code change suggestions.

Piinks reviewed (commented):
Thank you for all the context @TahaTesser!

@cruiser-baxter it looks like there are some failing tests here now, can you take a look?

`/b/s/w/ir/x/w/flutter/packages/flutter/test/material/slider_test.dart: Value indicator disappears after adjusting the slider`

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,56 @@ void main() {
+    // Value indicator is no longer visible.
+    expect(
+      valueIndicatorBox,
+      isNot(paints..path(color: theme.colorScheme.primary)..paragraph()),
```
```suggestion
      isNot(paints..scale()..path(color: theme.colorScheme.primary)),
```

TahaTesser commented on packages/flutter/test/material/slider_test.dart:
```diff
@@ -3507,6 +3507,56 @@ void main() {
+    valueIndicatorBox = tester.renderObject(find.byType(Overlay));
+    expect(
+      valueIndicatorBox,
+      paints..path(color: theme.colorScheme.primary)..paragraph(),
```
```suggestion
      paints..scale()..path(color: theme.colorScheme.primary),
```

TahaTesser reviewed (changes requested):
Just tried the test locally. After updating the test for M3 the expected order for `path` has changed.

TahaTesser committed:
Update packages/flutter/test/material/slider_test.dart

TahaTesser committed:
Update packages/flutter/test/material/slider_test.dart

TahaTesser committed:
Update packages/flutter/test/material/slider_test.dart

TahaTesser reviewed (approved):
LGTM

cruiser-baxter commented:
@TahaTesser Thank you for fixing that. I am in the middle of moving now and won't be able to look at this much for the next week.

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/169392212-02bb74fe-c6b9-4eb7-aa62-6ddb43a70d5b.png)
Thank you!
